Core: Add `BIND` macro for strongly typed enums

Kimau opened the pull request:
Currently, the existing BIND_ENUM_CONSTANT macro relies on implicit
conversion to integers. This works for standard C-style enums but
fails to compile with C++11 scoped enums (enum class), which require
explicit casting.

This commit adds a static_cast to int64_t, enabling support for 
modern, strictly typed C++ enums.
You want to use the modern C++11 enums to enable strict typing and 
conform memory layouts.

NOTE: This change must be implemented in core/object/class_db.h as it
modifies the base binding macros. It cannot be implemented as an addon
or GDExtension.

--- Example ---
```cpp
enum MyState { IDLE, RUN, WALK };
BIND_ENUM_CONSTANT(IDLE); // Works

enum class MyStateClass : uint8_t { IDLE, RUN, WALK };
BIND_ENUM_CONSTANT(MyStateClass::IDLE); // Now Works!
```

AThousandShips commented on core/object/class_db.h:
```diff
@@ -553,6 +553,9 @@ class ClassDB {
 	::ClassDB::bind_integer_constant(get_class_static(), __constant_get_enum_name(m_constant), #m_constant, m_constant);
 
+#define BIND_ENUM_CLASS_CONSTANT(m_constant) \
+	::ClassDB::bind_integer_constant( get_class_static(), _constant_get_enum_name(m_constant, #m_constant), #m_constant, static_cast<int64_t>(m_constant));
```
```suggestion
	::ClassDB::bind_integer_constant(get_class_static(), _constant_get_enum_name(m_constant, #m_constant), #m_constant, static_cast<int64_t>(m_constant));
```

dsnopek commented:
The overall idea of allowing the binding of enum classes makes sense to me! But I agree with @Ivorforce that we should update the original macro, rather than adding a new one

Kimau committed:
core: Updated BIND for strongly typed enums

Currently, the existing BIND_ENUM_CONSTANT macro relies on implicit
conversion to integers. This works for standard C-style enums but
fails to compile with C++11 scoped enums (enum class), which require
explicit casting.

This commit adds a static_cast to int64_t, enabling support for
modern, strictly typed C++ enums.
You want to use the modern C++11 enums to enable strict typing and
conform memory layouts.

NOTE: This change must be implemented in core/object/class_db.h as it
modifies the base binding macros. It cannot be implemented as an addon
or GDExtension.

--- Example ---
```cpp
enum MyState { IDLE, RUN, WALK };
BIND_ENUM_CONSTANT(IDLE); // Works

enum class MyStateClass : uint8_t { IDLE, RUN, WALK };
BIND_ENUM_CONSTANT(MyStateClass::IDLE); // Now Works!
```

Ivorforce reviewed (approved):
LGTM
We've been starting to prefer `enum class` over `enum`, so this is a step in the right direction.

The only caveat i can think of would be trying to bind `uint64_t` with values > `9.22337204e18`, but surely this won't be needed for enums 😄

dsnopek reviewed (approved):
Thanks, Looks good to me!

Repiteo reviewed (changes requested):
Unfortunately, we cannot simply reuse the old macro; at least, not without further adjustments. While this correctly binds, documentation is mangled because a class/namespace is now prefixed. For instance, even without changing an existing enum like `MouseMode` to a scoped equivalent, prefixing the class will still break that output:
```cpp
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_VISIBLE);
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_HIDDEN);
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_CAPTURED);
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_CONFINED);
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_CONFINED_HIDDEN);
	BIND_ENUM_CONSTANT(MouseMode::MOUSE_MODE_MAX);
```
```diff
 	<constants>
-		<constant name="MOUSE_MODE_VISIBLE" value="0" enum="MouseMode">
-			Makes the mouse cursor visible if it is hidden.
+		<constant name="MouseMode::MOUSE_MODE_VISIBLE" value="0" enum="MouseMode">
 		</constant>
-		<constant name="MOUSE_MODE_HIDDEN" value="1" enum="MouseMode">
-			Makes the mouse cursor hidden if it is visible.
+		<constant name="MouseMode::MOUSE_MODE_HIDDEN" value="1" enum="MouseMode">
 		</constant>
-		<constant name="MOUSE_MODE_CAPTURED" value="2" enum="MouseMode">
-			Captures the mouse. The mouse will be hidden and its position locked at the center of the window manager's window.
-			[b]Note:[/b] If you want to process the mouse's movement in this mode, you need to use [member InputEventMouseMotion.relative].
+		<constant name="MouseMode::MOUSE_MODE_CAPTURED" value="2" enum="MouseMode">
 		</constant>
-		<constant name="MOUSE_MODE_CONFINED" value="3" enum="MouseMode">
-			Confines the mouse cursor to the game window, and make it visible.
+		<constant name="MouseMode::MOUSE_MODE_CONFINED" value="3" enum="MouseMode">
 		</constant>
-		<constant name="MOUSE_MODE_CONFINED_HIDDEN" value="4" enum="MouseMode">
-			Confines the mouse cursor to the game window, and make it hidden.
+		<constant name="MouseMode::MOUSE_MODE_CONFINED_HIDDEN" value="4" enum="MouseMode">
 		</constant>
-		<constant name="MOUSE_MODE_MAX" value="5" enum="MouseMode">
-			Max value of the [enum MouseMode].
+		<constant name="MouseMode::MOUSE_MODE_MAX" value="5" enum="MouseMode">
 		</constant>
```
---
We *do* already have macros for handling enum classes, but those are currently isolated to `core_constants.cpp`. Specifically: `BIND_CORE_ENUM_CLASS_CONSTANT` and its variants.
```cpp
#define BIND_CORE_ENUM_CLASS_CONSTANT(m_enum, m_prefix, m_member)                                                  \
	{                                                                                                              \
		StringName enum_name = __constant_get_enum_name(m_enum::m_member);                                         \
		_global_constants.push_back(_CoreConstant(enum_name, #m_prefix "_" #m_member, (int64_t)m_enum::m_member)); \
		_global_constants_map[#m_prefix "_" #m_member] = _global_constants.size() - 1;                             \
		_global_enums[enum_name].push_back((_global_constants.ptr())[_global_constants.size() - 1]);               \
	}
```

They do more than simply casting the value and trimming the prefix though. This is because scoped enums take advantage of the fact that their class scope is required, and thus their contents have much shorter names. So if `MouseMode` was truly a scoped enum, it'd necessitate the following change:
```cpp
	enum class MouseMode {
		VISIBLE,
		HIDDEN,
		CAPTURED,
		CONFINED,
		CONFINED_HIDDEN,
		MAX,
	};
```
Otherwise declared variables would be comically verbose. In the most extreme case, failing to do so would result in `MouseMode::MOUSE_MODE_CONFINED_HIDDEN`.

This is a totally acceptable change, though! That's because the nature of using a scoped enum means that a ton of code will need to be touched anyway to account for explicit conversions and mandatory prefixes. In that same pass, there's no reason *not* to adjust the value names as well.

If this was a core enum type, we'd pass it to the macro which handles all this noise automatically:
```cpp
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, VISIBLE);
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, HIDDEN);
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, CAPTURED);
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, CONFINED);
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, CONFINED_HIDDEN);
	BIND_CORE_ENUM_CLASS_CONSTANT(MouseMode, MOUSE_MODE, MAX);
```
Everything outside of core has no such luxury. As such, I'd rather not integrate this change, as it'd hide a footgun of our own design. We absolutely *should* be able to support enum class binds out-of-the-box, but it needs to be accompanied with proper support for documentation
